daml-assistant/daml-helper/src/DamlHelper.hs

```diff
@@ -126,18 +130,227 @@ withJar jarPath args a = do
+--
+-- 2. If the target folder does not exist, it will error out and ask
+-- the user if they meant to use daml new instead.
+--
+-- 3. If the target folder is a daml project root, it will do nothing
+-- and let the user know the target is already a daml project.
+--
+-- 4. If the target folder is inside a daml project (transitively) but
+-- is not the project root, it will do nothing and print out a warning.
+--
+-- 5. If the target folder is a da project root, it will create a
+-- daml.yaml config file from the da.yaml config file, and let the
+-- user know that it did that.
+--
+-- 6. If the target folder is inside a da project (transitively) but
+-- is not the project root, it will error out with a message that lets
+-- the user know what the project root is and suggests the user run
+-- daml init on the project root.
+--
+-- 7. If none of the above, it will create a daml.yaml from scratch.
+-- It will attempt to find a Main.daml source file in the project
+-- directory tree, but if it does not it will use daml/Main.daml
+-- as the default.
+--
+runInit :: Maybe FilePath -> IO ()
+runInit targetFolderM = do
+    currentDir <- getCurrentDirectory
+    let targetFolder = fromMaybe currentDir targetFolderM
+        targetFolderRel = makeRelative currentDir targetFolder
+        projectConfigRel = normalise (targetFolderRel </> projectConfigName)
+          -- ^ for display purposes
+
+    -- cases 1 or 2
+    unlessM (doesDirectoryExist targetFolder) $ do
+        whenM (doesFileExist targetFolder) $ do
+            hPutStr stderr $ unlines
+                [ "ERROR: daml init target should be a directory, but is a file."
+                , "    target = " <> targetFolderRel
+                ]
+            exitFailure
+
+        hPutStr stderr $ unlines
+            [ "ERROR: daml init target does not exist."
+            , "    target = " <> targetFolderRel
+            , ""
+            , "To create a project directory use daml new instead:"
+            , "    daml new " <> escapePath targetFolderRel
+            ]
+        exitFailure
+    targetFolderAbs <- makeAbsolute targetFolder -- necessary to find project roots
+
+    -- cases 3 or 4
+    damlProjectRootM <- findDamlProjectRoot targetFolderAbs
+    whenJust damlProjectRootM $ \projectRoot -> do
+        let projectRootRel = makeRelative currentDir projectRoot
+        hPutStrLn stderr $ "DAML project already initialized at " <> projectRootRel
+        when (targetFolderAbs /= projectRoot) $ do
+            hPutStr stderr $ unlines
+                [ "WARNING: daml init target is not the DAML project root."
+                , "    daml init target  = " <> targetFolder
+                , "    DAML project root = " <> projectRootRel
+                ]
+        exitSuccess
+
+    -- cases 5 or 6
+    daProjectRootM <- findDaProjectRoot targetFolderAbs
+    whenJust daProjectRootM $ \projectRoot -> do
+        when (targetFolderAbs /= projectRoot) $ do
```

Comment:
Can we get a function `upgradeOldProject` and split this logic into a separate individual piece? I imagine you're going to want to write a couple of tests for that piece.